libnetwork/iptables/iptables.go

```diff
@@ -655,3 +696,256 @@ func (iptable IPTable) EnsureJumpRule(fromChain, toChain string) error {
+func (iptable IPTable) AddJumpRuleForIP(table Table, fromChain, toChain, ipaddr string) {
+	err := iptable.RawCombinedOutputNative("-t", string(table), "-C", fromChain, "-d", ipaddr, "-j", toChain)
+	if err == nil {
+		iptable.RawCombinedOutputNative("-t", string(table), "-F", toChain)
+	} else {
+		iptable.RawCombinedOutputNative("-t", string(table), "-N", toChain)
+		iptable.RawCombinedOutputNative("-t", string(table), "-I", fromChain, "-d", ipaddr, "-j", toChain)
+	}
+}
+
+//AddDNAT adds a dnat rule with a port
+func (iptable IPTable) AddDNATwithPort(table Table, chain, dstIP, proto, dstPort, natIP string) {
+	rule := []string{"-t", string(table), "-I", chain, "-d", dstIP, "-p", proto, "--dport", dstPort, "-j", "DNAT", "--to-destination", natIP}
+
+	if err := iptable.RawCombinedOutputNative(rule...); err != nil {
+		logrus.Errorf("set up rule failed, %v", err)
+	}
+}
+
+//AddSNAT adds a snat rule with a port
+func (iptable IPTable) AddSNATwithPort(table Table, chain, srcIP, proto, srcPort, natPort string) {
+	rule := []string{"-t", string(table), "-I", chain, "-s", srcIP, "-p", proto, "--sport", srcPort, "-j", "SNAT", "--to-source", ":" + natPort}
+	if err := iptable.RawCombinedOutputNative(rule...); err != nil {
+		logrus.Errorf("set up rule failed, %v", err)
+	}
+}
+
+//AddRedirect adds a redirect rule with a port
+func (iptable IPTable) AddRedirect(table Table, chain, dstIP, proto, dstPort, natIP string) {
+	rule := []string{"-t", string(table), "-I", chain, "-d", dstIP, "-p", proto, "--dport", dstPort, "-j", "REDIRECT", "--to-destination", natIP}
+
+	if err := iptable.RawCombinedOutputNative(rule...); err != nil {
+		logrus.Errorf("set up rule failed, %v", err)
+	}
+}
+
+//AddDropIncoming
+func (iptable IPTable) DropIncoming(table Table, chain, iface, addr string) error {
+	rule := []string{"-t", string(table), "-i", iface, "!", "-d", addr, "-j", "DROP"}
+	err := iptable.RawCombinedOutputNative(rule...)
+
+	if err != nil {
+		logrus.Errorf("set up rule failed, %v", err)
+	}
+	return err
+}
+
+//AddDropOutgoing
+func (iptable IPTable) DropOutgoing(table Table, chain, iface, addr string) error {
+	rule := []string{"-t", string(table), "-o", iface, "!", "-s", addr, "-j", "DROP"}
+	err := iptable.RawCombinedOutputNative(rule...)
+
+	if err != nil {
+		logrus.Errorf("set up rule failed, %v", err)
+	}
+	return err
+}
+
+func (iptable IPTable) GetInsertAction() string {
+	return string(Insert)
+}
+
+func (iptable IPTable) GetAppendAction() string {
+	return string(Append)
+}
+
+func (iptable IPTable) GetDeleteAction() string {
+	return string(Delete)
+}
+
+func (iptable IPTable) GetDropPolicy() string {
+	return string(Drop)
+}
+
+func (iptable IPTable) GetAcceptPolicy() string {
+	return string(Accept)
+}
+
+//Getters and setters for struct fields now that it's an interface
+func (c ChainInfo) GetName() string {
+	return c.Name
+}
+
+func (c ChainInfo) GetTable() Table {
+	return c.Table
+}
+
+func (c ChainInfo) SetTable(t Table) {
```

Comment:
As I'm trying to play around iptables/nftables on top of this PR, I tried to fix tests on my side. The new commit you added few hours ago is pretty much the same as what I did. However, the outstanding difference is here: 
this method won't work as its receiver isn't a pointer. Since `SetTable` is only used in this file, I think you should revert to the old code (ie. replace `SetTable` call sites by direct `c.Table` assignations). Same comment apply to nftables :wink:

Reply:
@evol262 did you saw this? This PR would potentially fix also #39590 #43755  #43870 so it would be nice to get it forward.